// hn-react/lib/components/deprecated/Paragraphs.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _EntityListMapper = require('../EntityListMapper');

var _EntityListMapper2 = _interopRequireDefault(_EntityListMapper);

var _utilDeprecate = require('util-deprecate');

var _utilDeprecate2 = _interopRequireDefault(_utilDeprecate);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Paragraphs(_ref) {
    var mapper = _ref.mapper,
        paragraphs = _ref.paragraphs,
        page = _ref.page,
        Wrapper = _ref.Wrapper,
        paragraphProps = _ref.paragraphProps;

    return _react2.default.createElement(_EntityListMapper2.default, (0, _assign2.default)({ entities: paragraphs, mapper: mapper, entityProps: (0, _assign2.default)({ page: page }, paragraphProps) }, Wrapper ? { entityWrapper: Wrapper } : {}));
}
Paragraphs.propTypes = {
    Wrapper: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.func]),
    mapper: _propTypes2.default.oneOfType([_propTypes2.default.shape(), _propTypes2.default.func]).isRequired,
    page: _propTypes2.default.shape(),
    paragraphProps: _propTypes2.default.shape(),
    paragraphs: _propTypes2.default.arrayOf(_propTypes2.default.shape())
};
Paragraphs.defaultProps = {
    Wrapper: undefined,
    page: undefined,
    paragraphProps: {}
};
exports.default = (0, _utilDeprecate2.default)(Paragraphs, 'The Paragraphs component is deprecated, use the EntityListMapper component instead');
