// hn-react/lib/components/deprecated/Paragraph.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _EntityMapper = require('../EntityMapper');

var _EntityMapper2 = _interopRequireDefault(_EntityMapper);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _utilDeprecate = require('util-deprecate');

var _utilDeprecate2 = _interopRequireDefault(_utilDeprecate);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Paragraph(_ref) {
    var mapper = _ref.mapper,
        uuid = _ref.uuid,
        page = _ref.page,
        index = _ref.index,
        paragraphProps = _ref.paragraphProps;

    return _react2.default.createElement(_EntityMapper2.default, { mapper: mapper, uuid: uuid, entityProps: (0, _assign2.default)({ index: index, page: page }, paragraphProps) });
}
Paragraph.propTypes = {
    index: _propTypes2.default.number,
    mapper: _propTypes2.default.oneOfType([_propTypes2.default.shape(), _propTypes2.default.func]).isRequired,
    page: _propTypes2.default.shape({}),
    paragraphProps: _propTypes2.default.shape(),
    uuid: _propTypes2.default.string.isRequired
};
Paragraph.defaultProps = {
    index: 0,
    page: undefined,
    paragraphProps: {}
};
exports.default = (0, _utilDeprecate2.default)(Paragraph, 'The Paragraph component is deprecated, use the EntityMapper component instead');
