// hn-react/lib/utils/tests/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.hnData = exports.entity = exports.uuid = exports.asyncMapper = exports.mapper = undefined;

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _hnData;

exports.mockSite = mockSite;

var _MapperComponent = require('./components/MapperComponent');

var _MapperComponent2 = _interopRequireDefault(_MapperComponent);

var _MapperComponentWithHOC = require('./components/MapperComponentWithHOC');

var _MapperComponentWithHOC2 = _interopRequireDefault(_MapperComponentWithHOC);

var _MapperComponentWithHOC3 = require('./components/MapperComponentWithHOC2');

var _MapperComponentWithHOC4 = _interopRequireDefault(_MapperComponentWithHOC3);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var __awaiter = undefined && undefined.__awaiter || function (thisArg, _arguments, P, generator) {
    return new (P || (P = _promise2.default))(function (resolve, reject) {
        function fulfilled(value) {
            try {
                step(generator.next(value));
            } catch (e) {
                reject(e);
            }
        }
        function rejected(value) {
            try {
                step(generator["throw"](value));
            } catch (e) {
                reject(e);
            }
        }
        function step(result) {
            result.done ? resolve(result.value) : new P(function (resolve) {
                resolve(result.value);
            }).then(fulfilled, rejected);
        }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};

var mapperComponent = jest.fn(_MapperComponent2.default);
var mapperComponentWithHOC = jest.fn(_MapperComponentWithHOC2.default);
var mapperComponentWithHOC2 = jest.fn(_MapperComponentWithHOC4.default);
var mapper = exports.mapper = {
    unique_type_1__unique_bundle_1: mapperComponent,
    unique_type_2__unique_bundle_2: mapperComponentWithHOC,
    unique_type_3__unique_bundle_3: mapperComponentWithHOC2
};
/* tslint:disable */
var asyncMapper = exports.asyncMapper = {
    unique_type_1__unique_bundle_1: jest.fn(function () {
        return _promise2.default.resolve(_MapperComponent2.default);
    }),
    unique_type_2__unique_bundle_2: jest.fn(function () {
        return _promise2.default.resolve(_MapperComponentWithHOC2.default);
    }),
    unique_type_3__unique_bundle_3: jest.fn(function () {
        return _promise2.default.resolve(_MapperComponentWithHOC4.default);
    })
};
/* tslint:enable */
var uuid = exports.uuid = 'unique-uuid-1';
var entity = exports.entity = {
    __hn: {
        entity: {
            bundle: 'unique_bundle_1',
            type: 'unique_type_1'
        }
    }
};
var entity2 = {
    __hn: {
        entity: {
            bundle: 'unique_bundle_2',
            type: 'unique_type_2'
        }
    }
};
var entity3 = {
    __hn: {
        entity: {
            bundle: 'unique_bundle_3',
            type: 'unique_type_3'
        }
    }
};
var hnData = exports.hnData = (_hnData = {}, (0, _defineProperty3.default)(_hnData, uuid, entity), (0, _defineProperty3.default)(_hnData, 'unique-uuid-2', entity), (0, _defineProperty3.default)(_hnData, 'unique-uuid-3', entity2), (0, _defineProperty3.default)(_hnData, 'unique-uuid-4', entity3), _hnData);
function mockSite() {
    var SiteMock = function SiteMock() {
        var _this = this;

        (0, _classCallCheck3.default)(this, SiteMock);

        this.getData = jest.fn(function (u) {
            return hnData[u];
        });
        this.getPage = jest.fn(function () {
            return __awaiter(_this, void 0, void 0, /*#__PURE__*/_regenerator2.default.mark(function _callee() {
                return _regenerator2.default.wrap(function _callee$(_context) {
                    while (1) {
                        switch (_context.prev = _context.next) {
                            case 0:
                                return _context.abrupt('return', uuid);

                            case 1:
                            case 'end':
                                return _context.stop();
                        }
                    }
                }, _callee, this);
            }));
        });
    };

    return new SiteMock();
}
