// hn-react/lib/components/EntityListMapper.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _EntityMapper = require('./EntityMapper');

var _EntityMapper2 = _interopRequireDefault(_EntityMapper);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var EntityListMapper = function EntityListMapper(_ref) {
    var mapper = _ref.mapper,
        entities = _ref.entities,
        entityWrapper = _ref.entityWrapper,
        entityProps = _ref.entityProps,
        asyncMapper = _ref.asyncMapper;
    return entities.map(function (ref, index) {
        var EntityWrapper = entityWrapper || _react.Fragment;
        var uuid = ref.target_uuid || ref;
        return _react2.default.createElement(EntityWrapper, { key: uuid }, _react2.default.createElement(_EntityMapper2.default, { uuid: uuid, mapper: mapper, asyncMapper: asyncMapper, entityProps: (0, _assign2.default)({}, entityProps, { index: index }) }));
    });
};
EntityListMapper.propTypes = {
    asyncMapper: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.oneOfType([_propTypes2.default.shape(), _propTypes2.default.func])]),
    entities: _propTypes2.default.arrayOf(_propTypes2.default.oneOfType([_propTypes2.default.shape({
        target_uuid: _propTypes2.default.string.isRequired
    }), _propTypes2.default.string])).isRequired,
    entityProps: _propTypes2.default.shape(),
    entityWrapper: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.func]),
    mapper: _propTypes2.default.oneOfType([_propTypes2.default.shape(), _propTypes2.default.func])
};
EntityListMapper.defaultProps = {
    asyncMapper: undefined,
    entityProps: undefined,
    entityWrapper: undefined,
    mapper: undefined
};
exports.default = EntityListMapper;
